File: rust/cmd/minimax/src/commands/music.rs
//! Music generation commands.
//!
//! Compatible with Go version's music commands.

use clap::{Args, Subcommand};

use giztoy_minimax::{MusicRequest, MODEL_MUSIC_20};

use super::{
    create_client, format_bytes, get_context, load_request, output_bytes, output_result,
    print_success, print_verbose, require_input_file,
};
use crate::Cli;

/// Music generation service.
///
/// Supports music generation from prompts and lyrics.
#[derive(Args)]
pub struct MusicCommand {
    #[command(subcommand)]
    command: MusicSubcommand,
}

#[derive(Subcommand)]
enum MusicSubcommand {
    /// Generate music from prompt and lyrics
    Generate,
}

impl MusicCommand {
    pub async fn run(&self, cli: &Cli) -> anyhow::Result<()> {
        match &self.command {
            MusicSubcommand::Generate => self.generate(cli).await,
        }
    }

    async fn generate(&self, cli: &Cli) -> anyhow::Result<()> {
        let input_file = require_input_file(cli)?;
        let ctx = get_context(cli)?;

        let mut req: MusicRequest = load_request(input_file)?;

        // Use defaults if not specified
        if req.model.is_none() {
            req.model = Some(MODEL_MUSIC_20.to_string());
        }

        print_verbose(cli, &format!("Using context: {}", ctx.name));
        print_verbose(
            cli,
            &format!("Model: {}", req.model.as_deref().unwrap_or("default")),
        );

        let client = create_client(&ctx)?;
        let resp = client.music().generate(&req).await?;

        // Determine output file extension
        let ext = req.format.as_deref().unwrap_or("mp3");
        let default_output = format!("music.{}", ext);
        let output_path = cli.output.as_deref().unwrap_or(&default_output);

        // Write audio to file
        output_bytes(&resp.audio, output_path)?;
        print_success(&format!(
            "Music saved to: {} ({})",
            output_path,
            format_bytes(resp.audio.len())
        ));

        // Output final info
        let result = serde_json::json!({
            "audio_size": resp.audio.len(),
            "duration": resp.duration,
            "extra_info": resp.extra_info,
            "output_file": output_path,
        });

        output_result(&result, None, cli.json)
    }
}

File: go/cmd/minimax/commands/music.go
package commands

import (
	"context"
	"fmt"
	"time"

	"github.com/spf13/cobra"

	"github.com/haivivi/giztoy/go/pkg/minimax"
)

var musicCmd = &cobra.Command{
	Use:   "music",
	Short: "Music generation service",
	Long: `Music generation service.

Generate music with vocals based on prompts and lyrics.

Example request file (music.yaml):
  model: music-2.0
  prompt: Pop music, upbeat, suitable for summer
  lyrics: |
    [Verse]
    Walking down the street
    Feeling the summer heat
    [Chorus]
    Life is beautiful
    Every day is wonderful
  sample_rate: 44100
  format: mp3`,
}

var musicGenerateCmd = &cobra.Command{
	Use:   "generate",
	Short: "Generate music",
	Long: `Generate music with vocals.

Examples:
  minimax -c myctx music generate -f music.yaml -o song.mp3
  minimax -c myctx music generate -f music.yaml --json`,
	RunE: func(cmd *cobra.Command, args []string) error {
		if err := requireInputFile(); err != nil {
			return err
		}

		ctx, err := getContext()
		if err != nil {
			return err
		}

		var req minimax.MusicRequest
		if err := loadRequest(getInputFile(), &req); err != nil {
			return err
		}

		if req.Model == "" {
			req.Model = minimax.ModelMusic20
		}

		printVerbose("Using context: %s", ctx.Name)
		printVerbose("Model: %s", req.Model)
		printVerbose("Prompt: %s", req.Prompt)

		// Create API client
		client := createClient(ctx)

		// Call API - music generation can take a while
		reqCtx, cancel := context.WithTimeout(context.Background(), 300*time.Second)
		defer cancel()

		resp, err := client.Music.Generate(reqCtx, &req)
		if err != nil {
			return fmt.Errorf("music generation failed: %w", err)
		}

		// Output audio to file if specified
		outputPath := getOutputFile()
		if outputPath != "" && len(resp.Audio) > 0 {
			if err := outputBytes(resp.Audio, outputPath); err != nil {
				return fmt.Errorf("failed to write audio file: %w", err)
			}
			printSuccess("Music saved to: %s (%s)", outputPath, formatBytes(len(resp.Audio)))
		}

		// Output result
		result := map[string]any{
			"audio_size":  len(resp.Audio),
			"duration_ms": resp.Duration,
			"extra_info":  resp.ExtraInfo,
			"output_file": outputPath,
		}

		return outputResult(result, "", isJSONOutput())
	},
}

func init() {
	musicCmd.AddCommand(musicGenerateCmd)
}
